### DISCUSSION TITLE
v2.0 Planning

### DISCUSSION THREAD

**Discussion by jshlbrd** [brexhq/substation#224] [General]
@shellcromancer and I were discussing breaking changes in #215, and we've accumulated some that are worth mentioning and preparing for a small v2.0 release. This includes:
- Splitting the libsonnet file and bringing part of it in scope to SemVer (discussed in #215)
- Refactoring the libsonnet functions (discussed in #215)
- Removing `Transform` fields in all `Meta*` transforms (replaced by `Transforms`)
- Removing the `MetaPipeline` transform (replaced by the change above)
- Changing the default behavior of the `MetaErr` transform (empty err list will not catch every error)

Up for possible inclusion is removing retry functionality from each individual transform (replaced by `MetaRetry` transform).

**Comment by jshlbrd** [comment]
I started a [v2 branch](https://github.com/brexhq/substation/tree/v2/v2) for the Brex team to work on. We should add comments to this discussion for any major changes and discuss them in a thread.

**Comment by jshlbrd** [comment]
# Refactoring the Condition Package

[This commit](https://github.com/brexhq/substation/commit/433a599de929cd332180e0d2d64a07d75a27d4ba) is the start of a big change, but shouldn't have much impact on v1.0 deployments. In v1.0 and previous releases the condition package uses a [config schema](https://github.com/brexhq/substation/blob/main/condition/condition.go#L17) that separates operators (any, all, none) and inspectors, which required the addition of the [meta_condition inspector](https://github.com/brexhq/substation/blob/main/condition/meta_condition.go) for complex logic (chaining ORs and ANDs or ANDs and ORs). This change simplifies the config schema so that v1.0 operators are now inspectors, which makes them easier to chain.

This is required in v1.0:
```jsonnet

sub.cnd.all([
  sub.cnd.meta.condition({ condition: sub.cnd.any([
    sub.cnd.str.contains({value: "a"}),
    sub.cnd.str.contains({value: "b"}),
    sub.cnd.str.contains({value: "c"}),
  ]),
  sub.cnd.meta.condition({ condition: sub.cnd.none([
    sub.cnd.str.contains({value: "z"}),
  ]),
])
```

The equivalent in v2.0 would be:
```jsonnet

sub.cnd.all([
  sub.cnd.any([
    sub.cnd.str.contains({value: "a"}),
    sub.cnd.str.contains({value: "b"}),
    sub.cnd.str.contains({value: "c"}),
  ]),
  sub.cnd.none([
    sub.cnd.str.contains({value: "z"}),
  ]),
])
```

This also removes the need for the other meta inspectors:
- `meta_for_each`: This behavior is now built into `meta_all`, `meta_any`, `meta_none` when the input is an array.
- `meta_negate`: This is replaced by `meta_none`.
- `meta_err`: Inspectors no longer produce errors that should be caught at runtime, so this isn't needed.

The v2.0 libsonnet uses the same shortcut as the v1.0 operators, so the migration path is "replace all" in existing configs; for example, replacing `sub.cnd.meta.condition({ condition: sub.cnd.any([` with `sub.cnd.any([` should be all it takes to migrate. This type of migration should be tested before v2.0 releases to confirm assumptions about simplicity in migrating.

**Reply by jshlbrd** [reply]
@shellcromancer This should address some issues we discussed about the complexity of chaining conditions, but let me know what you think.

**Reply by shellcromancer** [reply]
The simplifications to remove `meta.condition` look great and more intuitive to me, thanks for kicking these changes off! 👍 

> meta_err: Inspectors no longer produce errors that should be caught at runtime, so this isn't needed.

Will `meta_err` still be needed for transforms that produce errors at runtime? One example is with the the `dns.ip_lookup` transform like used [here](https://github.com/shellcromancer/substation-ocsf/blob/main/patterns.libsonnet#L35).

**Reply by jshlbrd** [reply]
Yeah! We're keeping the `meta.err` transform, but not the inspector variant. 👍

**Reply by shellcromancer** [reply]
Gotcha. I got confused by the duplicative name -- no concerns in that case!

**Comment by jshlbrd** [comment]
The [v2.0.0 beta release](https://github.com/brexhq/substation/releases/tag/v2.0.0-beta) is up and the [PR](https://github.com/brexhq/substation/pull/228) to bump versions is drafted. 👍
